### ISSUE TITLE
SwiftUI sheet sometimes fails to present

### ISSUE THREAD

**Issue by aheze** [aheze/Popovers#11]
Sometimes when I try to present a sheet, nothing happens, and this is printed to the console:

> 2022-01-05 09:37:44.755224-0800 PopoversPlaygroundsApp[83666:1047669] [Presentation] Attempt to present <_TtGC7SwiftUI29PresentationHostingControllerVS_7AnyView_: 0x129811c00> on <_TtGC7SwiftUI19UIHostingControllerGVS_15ModifiedContentVS_7AnyViewVS_12RootModifier__: 0x127e0e060> (from <_TtGC7SwiftUI19UIHostingControllerGVS_15ModifiedContentVVS_22_VariadicView_Children7ElementVS_24NavigationColumnModifier__: 0x127e09d90>) which is already presenting <Popovers.PopoverContainerViewController: 0x12c70a5d0>.

**Comment by ShezHsky** [comment]
This is reproducible in the sample app if you tap on the "Context Menu" button, then tap on the "Popover from Sheet" button *before* dismissing the context menu. This makes sense given the popover's view controller hasn't been dismissed yet, so presenting a sheet on the view controller that is still presenting the popover view controller won't work.

One option could be to drop the use of a presented view controller and insert a transient subview in the window (brought to the front) instead, leaving the topmost view controller free for presenting anything else. This would also address #13

**Comment by aheze** [comment]
Should be an easy fix. I’ll try it out.

**Comment by aheze** [comment]
Edit: was not an easy fix, am stuck ;-;

**Comment by ShezHsky** [comment]
Apols for not poking sooner, bit of a wild start to the year here. I've [pushed a branch to my fork](https://github.com/ShezHsky/Popovers/tree/feature/embed-view-into-parent) that seems to be pretty close to desirable - swaps out presenting the view controller for shoving the container view into the window's view hierarchy. Focus isn't taken by the popover, but it does mean Voiceover won't read out the popover anymore. If this were UIKit I'd use `UIAccessibilityPostNotification` to tell Voiceover the layout has changed, and the popover view should take focus - not sure how to go about doing this in SwiftUI. Deffo needs more work but might be a good starting point?

**Comment by aheze** [comment]
@ShezHsky thanks for taking a look. Yeah this year has been pretty busy for me too. 

> shoving the container view into the window's view hierarchy

I tried doing this too, but couldn't add the popover container view controller's view in the window - some hierarchy error. So I omitted `addChild(childViewController)` and `childViewController.didMove(toParent: self)`, and then got some weird visual glitches where popovers wouldn't disappear. How did you get this to work without errors?

```swift
let hostingController = UIHostingController(rootView: popoverContainerView)
hostingController.view.frame = bounds
hostingController.view.backgroundColor = .clear
hostingController.view.autoresizingMask = [.flexibleWidth, .flexibleHeight]

addSubview(hostingController.view)
setNeedsLayout()
layoutIfNeeded()

onMovedToWindow?()
```

https://github.com/ShezHsky/Popovers/blob/d3adcb6ff9cfa0cdfa5238e29c0fab2b9a3b6ca3/Sources/PopoverGestureContainer.swift#L23

**Comment by ShezHsky** [comment]
> How did you get this to work without errors?

Is this in comparison to [this commit](https://github.com/aheze/Popovers/commit/afa2959b3de00c3be7bdb0cc9af2180c12e515e4)? Though tbh looking at the code/running it on a sim here it seems to also behave itself, so unless I'm missing something I can't see why the popover would linger after removing the container view from the view hierarchy.

**Comment by aheze** [comment]
> Is this in comparison to [this commit](https://github.com/aheze/Popovers/commit/afa2959b3de00c3be7bdb0cc9af2180c12e515e4)?

Yeah kind of. That commit is pretty stable, except overlaying sheets cover the popover. The errors I got were with 2 uncommitted changes that I tried:

1. Replace `topmostViewController.view` with `window`

```swift
topmostViewController.add(
    childViewController: newPopoverViewController,
    in: window
)
```

Result:
```
2022-01-16 17:29:11.529266-0800 PopoversXcodeApp[25717:1779371] [UIFocus] Focus system enabled
2022-01-16 17:29:37.037912-0800 PopoversXcodeApp[25717:1779371] *** Terminating app due to uncaught exception 'UIViewControllerHierarchyInconsistency', reason: 'child view controller:<Popovers.PopoverContainerViewController: 0x1445a2a60> should have parent view controller:(null) but actual parent is:<_TtGC7SwiftUI19UIHostingControllerGVS_15ModifiedContentVS_7AnyViewVS_12RootModifier__: 0x1447188c0>'
*** First throw call stack:
(
	0   CoreFoundation                      0x00000001803e1188 __exceptionPreprocess + 236
	1   libobjc.A.dylib                     0x0000000180193384 objc_exception_throw + 56
	2   CoreFoundation                      0x00000001803e1058 -[NSException initWithCoder:] + 0
	3   UIKitCore                           0x00000001852eb070 -[UIView(Hierarchy) _associatedViewControllerForwardsAppearanceCallbacks:performHierarchyCheck:isRoot:] + 284
	4   UIKitCore                           0x00000001852eb89c -[UIView(Hierarchy) _willMoveToWindow:withAncestorView:] + 244
	5   UIKitCore                           0x00000001852fb0d4 -[UIView(Internal) _addSubview:positioned:relativeTo:] + 432
	6   PopoversXcodeApp                    0x00000001043eb948 $sSo16UIViewControllerC8PopoversE3add09childViewB02inyAB_So0A0CtF + 248
	7   PopoversXcodeApp                    0x00000001043b5960 $s8Popo

...
```

2. Replace `topmostViewController.add(...)` with:

```swift
/// Add Child View as Subview
window.addSubview(newPopoverViewController.view)

/// Configure Child View
newPopoverViewController.view.frame = window.bounds
newPopoverViewController.view.autoresizingMask = [.flexibleWidth, .flexibleHeight]
```
This kind of work, but it seem like the popovers just keep stacking on top of each other:

https://user-images.githubusercontent.com/49819455/149687669-ab3ee86e-9859-4b02-b1fd-39a4d2961339.mov

Anyway I'll check out your commit and get back to you in a bit.

**Comment by aheze** [comment]
Yo @ShezHsky nice work! I only found 1 bug where resizing windows doesn't update the container. I'll try to fix

![Screen Shot 2022-01-16 at 7 35 37 PM](https://user-images.githubusercontent.com/49819455/149704239-c3ddfb0f-58a9-44f2-87c2-aeb51db29f19.png)

**Comment by aheze** [comment]
Finally, an easy fix :D

![Screen Shot 2022-01-16 at 7 43 41 PM](https://user-images.githubusercontent.com/49819455/149704839-9754c859-0987-4bc4-aef0-33f0e5113c73.png)

**Comment by aheze** [comment]
This seems to work ok in making voiceover focus the popover:
```swift
UIAccessibility.post(notification: .screenChanged, argument: nil)
```

**Comment by aheze** [comment]
A problem is tapping outside doesn't work with VoiceOver since [`hitTest` isn't called](https://stackoverflow.com/a/48532092/14351818), so I think the only way is something like this (where each red view is a button):

![Accessibility Dismiss](https://user-images.githubusercontent.com/49819455/149707879-d6ee8ece-d22a-4580-b2f7-d98dac1510f7.png)

This is entirely possible but just thinking about it, I don't want to code it. Maybe a button instead?

![Screen Shot 2022-01-16 at 8 26 42 PM](https://user-images.githubusercontent.com/49819455/149708056-167f1ab1-7624-4c8e-97df-b4ddbd9373ba.png)

I could check `UIAccessibility.isVoiceOverRunning` and only show the button when it's true. The client can always add their own button inside a popover, but what is it's a tip or something? Seems a bit clunky. I could add a property inside `Popover.Attributes` to enable/disable auto voiceover button.

**Comment by aheze** [comment]
@ShezHsky I've added the VoiceOver support in a PR to your fork: https://github.com/ShezHsky/Popovers/pull/2.

![Popovers automatically have an X button](https://user-images.githubusercontent.com/49819455/149715492-244edffb-3c27-4f90-9f0a-8db9c28e051d.png)
![The X button is customizable](https://user-images.githubusercontent.com/49819455/149715494-61039b2f-bb5a-46b1-9963-4a1524671194.png)

**Comment by ShezHsky** [comment]
Nice one, cheers - I've raised #15 that contains those changes + the accessibility escape gesture. There's some extra notes in there re another bit that might need a look, however I won't have time tonight to have a look (in case you fancy a check). Otherwise will come back to this tomorrow

**Comment by aheze** [comment]
Did some polishing up - fixed in [1.2.0](https://github.com/aheze/Popovers/releases/tag/1.2.0).
